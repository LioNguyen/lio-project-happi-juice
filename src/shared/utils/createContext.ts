import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
