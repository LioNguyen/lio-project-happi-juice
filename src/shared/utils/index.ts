export * from './lazyLoad'
export * from './tailwind'
