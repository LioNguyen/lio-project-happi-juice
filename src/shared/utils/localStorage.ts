/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
