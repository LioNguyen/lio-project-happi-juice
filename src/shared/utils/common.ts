import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
