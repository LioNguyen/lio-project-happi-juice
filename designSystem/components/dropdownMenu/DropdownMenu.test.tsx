import { screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import {
  DropdownMenu,
  DropdownMenuCheckboxItem,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuRadioGroup,
  DropdownMenuRadioItem,
  DropdownMenuSeparator,
  DropdownMenuShortcut,
  DropdownMenuSub,
  DropdownMenuSubContent,
  DropdownMenuSubTrigger,
  DropdownMenuTrigger,
} from ".";

// Mock lucide-react icons
vi.mock("lucide-react", () => ({
  Check: () => <div data-testid="check-icon" />,
  ChevronRight: () => <div data-testid="chevron-right-icon" />,
  Circle: () => <div data-testid="circle-icon" />,
}));

// Mock radix-ui dropdown menu components
vi.mock("@radix-ui/react-dropdown-menu", () => ({
  Root: vi.fn(({ children }) => <div data-testid="dropdown-root">{children}</div>),
  Trigger: vi.fn(({ children }) => <button data-testid="dropdown-trigger">{children}</button>),
  Portal: vi.fn(({ children }) => <div data-testid="dropdown-portal">{children}</div>),
  Content: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-content" className={className} {...props}>
      {children}
    </div>
  )),
  Item: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-item" className={className} {...props}>
      {children}
    </div>
  )),
  CheckboxItem: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-checkbox-item" className={className} {...props}>
      {children}
    </div>
  )),
  RadioItem: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-radio-item" className={className} {...props}>
      {children}
    </div>
  )),
  Label: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-label" className={className} {...props}>
      {children}
    </div>
  )),
  Separator: vi.fn(({ className, ...props }) => (
    <div data-testid="dropdown-separator" className={className} {...props} />
  )),
  Sub: vi.fn(({ children }) => <div data-testid="dropdown-sub">{children}</div>),
  SubTrigger: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-sub-trigger" className={className} {...props}>
      {children}
    </div>
  )),
  SubContent: vi.fn(({ className, children, ...props }) => (
    <div data-testid="dropdown-sub-content" className={className} {...props}>
      {children}
    </div>
  )),
  RadioGroup: vi.fn(({ children }) => <div data-testid="dropdown-radio-group">{children}</div>),
  Group: vi.fn(({ children }) => <div data-testid="dropdown-group">{children}</div>),
  ItemIndicator: vi.fn(({ children }) => <div data-testid="dropdown-item-indicator">{children}</div>),
}));

describe("DropdownMenu Component", () => {
  const renderDropdownMenu = (props = {}) => {
    return renderWithTheme(
      <DropdownMenu>
        <DropdownMenuTrigger>Open Menu</DropdownMenuTrigger>
        <DropdownMenuContent {...props}>
          <DropdownMenuLabel>My Account</DropdownMenuLabel>
          <DropdownMenuSeparator />
          <DropdownMenuItem>Profile</DropdownMenuItem>
          <DropdownMenuSub>
            <DropdownMenuSubTrigger>More Options</DropdownMenuSubTrigger>
            <DropdownMenuSubContent>
              <DropdownMenuItem>Sub Item</DropdownMenuItem>
            </DropdownMenuSubContent>
          </DropdownMenuSub>
          <DropdownMenuRadioGroup>
            <DropdownMenuRadioItem value="1">Option 1</DropdownMenuRadioItem>
            <DropdownMenuRadioItem value="2">Option 2</DropdownMenuRadioItem>
          </DropdownMenuRadioGroup>
          <DropdownMenuCheckboxItem checked>Show Status</DropdownMenuCheckboxItem>
        </DropdownMenuContent>
      </DropdownMenu>,
    );
  };

  // Basic Rendering Tests
  it("should render all dropdown menu components", () => {
    renderDropdownMenu();

    expect(screen.getByTestId("dropdown-root")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-trigger")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-content")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-label")).toBeInTheDocument();
    expect(screen.getByTestId("dropdown-separator")).toBeInTheDocument();
    expect(screen.getAllByTestId("dropdown-item")[0]).toBeInTheDocument();
  });

  // Snapshot Tests
  describe("DropdownMenu Snapshots", () => {
    it("should match snapshot with default props", () => {
      const { container } = renderDropdownMenu();
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with custom content", () => {
      const { container } = renderWithTheme(
        <DropdownMenu>
          <DropdownMenuTrigger>Custom Trigger</DropdownMenuTrigger>
          <DropdownMenuContent>
            <DropdownMenuItem>Custom Item</DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  // DropdownMenuContent Tests
  describe("DropdownMenuContent", () => {
    it("should render with correct classes", () => {
      renderDropdownMenu();

      const content = screen.getByTestId("dropdown-content");
      expect(content).toHaveClass(
        "z-50",
        "min-w-[8rem]",
        "overflow-hidden",
        "rounded-md",
        "border",
        "bg-popover",
        "p-1",
        "text-popover-foreground",
        "shadow-md",
      );
    });
  });

  // DropdownMenuItem Tests
  describe("DropdownMenuItem", () => {
    it("should render with correct classes", () => {
      renderDropdownMenu();

      const item = screen.getAllByTestId("dropdown-item");
      expect(item[0]).toHaveClass(
        "relative",
        "flex",
        "cursor-default",
        "select-none",
        "items-center",
        "rounded-sm",
        "px-2",
        "py-1.5",
        "text-sm",
        "outline-none",
        "focus:bg-secondary",
        "focus:text-text-primary",
      );
    });

    it("should render with inset class when inset prop is true", () => {
      renderWithTheme(
        <DropdownMenu>
          <DropdownMenuContent>
            <DropdownMenuItem inset>Inset Item</DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>,
      );

      const item = screen.getByTestId("dropdown-item");
      expect(item).toHaveClass("pl-8");
    });
  });

  // DropdownMenuCheckboxItem Tests
  describe("DropdownMenuCheckboxItem", () => {
    it("should render with correct classes and check icon when checked", () => {
      renderDropdownMenu();

      const checkboxItem = screen.getByTestId("dropdown-checkbox-item");
      expect(checkboxItem).toHaveClass(
        "relative",
        "flex",
        "cursor-default",
        "select-none",
        "items-center",
        "rounded-sm",
        "py-1.5",
        "pl-8",
        "pr-2",
        "text-sm",
        "focus:bg-secondary",
        "focus:text-text-primary",
      );
    });
  });

  // DropdownMenuRadioItem Tests
  describe("DropdownMenuRadioItem", () => {
    it("should render with correct classes", () => {
      renderDropdownMenu();

      const radioItems = screen.getAllByTestId("dropdown-radio-item");
      radioItems.forEach(item => {
        expect(item).toHaveClass(
          "relative",
          "flex",
          "cursor-default",
          "select-none",
          "items-center",
          "rounded-sm",
          "py-1.5",
          "pl-8",
          "pr-2",
          "text-sm",
        );
      });
    });
  });

  // DropdownMenuLabel Tests
  describe("DropdownMenuLabel", () => {
    it("should render with correct classes", () => {
      renderDropdownMenu();

      const label = screen.getByTestId("dropdown-label");
      expect(label).toHaveClass("px-2", "py-1.5", "text-sm", "font-semibold");
    });

    it("should render with inset class when inset prop is true", () => {
      renderWithTheme(
        <DropdownMenu>
          <DropdownMenuContent>
            <DropdownMenuLabel inset>Inset Label</DropdownMenuLabel>
          </DropdownMenuContent>
        </DropdownMenu>,
      );

      const label = screen.getByTestId("dropdown-label");
      expect(label).toHaveClass("pl-8");
    });
  });

  // DropdownMenuSeparator Tests
  describe("DropdownMenuSeparator", () => {
    it("should render with correct classes", () => {
      renderDropdownMenu();

      const separator = screen.getByTestId("dropdown-separator");
      expect(separator).toHaveClass("-mx-1", "my-1", "h-px", "bg-secondary");
    });
  });

  // DropdownMenuShortcut Tests
  describe("DropdownMenuShortcut", () => {
    it("should render with correct classes", () => {
      renderWithTheme(
        <DropdownMenu>
          <DropdownMenuContent>
            <DropdownMenuItem>
              Item
              <DropdownMenuShortcut>⌘K</DropdownMenuShortcut>
            </DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>,
      );

      const shortcut = screen.getByText("⌘K");
      expect(shortcut).toHaveClass("ml-auto", "text-xs", "tracking-widest", "opacity-60");
    });
  });

  // SubMenu Tests
  describe("SubMenu Components", () => {
    it("should render submenu components with correct classes", () => {
      renderWithTheme(
        <DropdownMenu>
          <DropdownMenuContent>
            <DropdownMenuSub>
              <DropdownMenuSubTrigger inset>More Options</DropdownMenuSubTrigger>
              <DropdownMenuSubContent>
                <DropdownMenuItem>Sub Item</DropdownMenuItem>
              </DropdownMenuSubContent>
            </DropdownMenuSub>
          </DropdownMenuContent>
        </DropdownMenu>,
      );

      const subTrigger = screen.getByTestId("dropdown-sub-trigger");
      expect(subTrigger).toHaveClass(
        "flex",
        "cursor-default",
        "select-none",
        "items-center",
        "rounded-sm",
        "px-2",
        "py-1.5",
        "text-sm",
        "outline-none",
      );
      expect(subTrigger).toHaveClass("pl-8");

      const subContent = screen.getByTestId("dropdown-sub-content");
      expect(subContent).toHaveClass(
        "z-50",
        "min-w-[8rem]",
        "overflow-hidden",
        "rounded-md",
        "border",
        "bg-popover",
        "p-1",
        "text-popover-foreground",
      );
    });
  });
});
