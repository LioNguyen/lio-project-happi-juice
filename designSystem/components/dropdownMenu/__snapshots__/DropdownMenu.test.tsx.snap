// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu Component > DropdownMenu Snapshots > should match snapshot with custom content 1`] = `
<div>
  <div
    data-testid="dropdown-root"
  >
    <button
      data-testid="dropdown-trigger"
    >
      Custom Trigger
    </button>
    <div
      data-testid="dropdown-portal"
    >
      <div
        class="z-50 min-w-[8rem] overflow-hidden rounded-md border bg-popover p-1 text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-testid="dropdown-content"
        sideoffset="4"
      >
        <div
          class="relative flex cursor-default select-none items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-secondary focus:text-text-primary data-[disabled]:pointer-events-none data-[disabled]:opacity-50 [&>svg]:size-4 [&>svg]:shrink-0"
          data-testid="dropdown-item"
        >
          Custom Item
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu Component > DropdownMenu Snapshots > should match snapshot with default props 1`] = `
<div>
  <div
    data-testid="dropdown-root"
  >
    <button
      data-testid="dropdown-trigger"
    >
      Open Menu
    </button>
    <div
      data-testid="dropdown-portal"
    >
      <div
        class="z-50 min-w-[8rem] overflow-hidden rounded-md border bg-popover p-1 text-popover-foreground shadow-md data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
        data-testid="dropdown-content"
        sideoffset="4"
      >
        <div
          class="px-2 py-1.5 text-sm font-semibold"
          data-testid="dropdown-label"
        >
          My Account
        </div>
        <div
          class="-mx-1 my-1 h-px bg-secondary"
          data-testid="dropdown-separator"
        />
        <div
          class="relative flex cursor-default select-none items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-secondary focus:text-text-primary data-[disabled]:pointer-events-none data-[disabled]:opacity-50 [&>svg]:size-4 [&>svg]:shrink-0"
          data-testid="dropdown-item"
        >
          Profile
        </div>
        <div
          data-testid="dropdown-sub"
        >
          <div
            class="flex cursor-default gap-2 select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none focus:bg-accent data-[state=open]:bg-accent [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0"
            data-testid="dropdown-sub-trigger"
          >
            More Options
            <div
              data-testid="chevron-right-icon"
            />
          </div>
          <div
            class="z-50 min-w-[8rem] overflow-hidden rounded-md border bg-popover p-1 text-popover-foreground shadow-lg data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
            data-testid="dropdown-sub-content"
          >
            <div
              class="relative flex cursor-default select-none items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-secondary focus:text-text-primary data-[disabled]:pointer-events-none data-[disabled]:opacity-50 [&>svg]:size-4 [&>svg]:shrink-0"
              data-testid="dropdown-item"
            >
              Sub Item
            </div>
          </div>
        </div>
        <div
          data-testid="dropdown-radio-group"
        >
          <div
            class="relative flex cursor-default select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
            data-testid="dropdown-radio-item"
            value="1"
          >
            <span
              class="absolute left-2 flex h-3.5 w-3.5 items-center justify-center"
            >
              <div
                data-testid="dropdown-item-indicator"
              >
                <div
                  data-testid="circle-icon"
                />
              </div>
            </span>
            Option 1
          </div>
          <div
            class="relative flex cursor-default select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
            data-testid="dropdown-radio-item"
            value="2"
          >
            <span
              class="absolute left-2 flex h-3.5 w-3.5 items-center justify-center"
            >
              <div
                data-testid="dropdown-item-indicator"
              >
                <div
                  data-testid="circle-icon"
                />
              </div>
            </span>
            Option 2
          </div>
        </div>
        <div
          class="relative flex cursor-default select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none transition-colors focus:bg-secondary focus:text-text-primary data-[disabled]:pointer-events-none data-[disabled]:opacity-50"
          data-testid="dropdown-checkbox-item"
        >
          <span
            class="absolute left-2 flex h-3.5 w-3.5 items-center justify-center"
          >
            <div
              data-testid="dropdown-item-indicator"
            >
              <div
                data-testid="check-icon"
              />
            </div>
          </span>
          Show Status
        </div>
      </div>
    </div>
  </div>
</div>
`;
